### ISSUE TITLE
Question regarding script object destruction

### ISSUE THREAD

**Issue by ErrorFlexXx** [cpgf/cpgf#73]
Hello,

first of all, thank you for that big library with many nice features.

Currently I build a program that exposes some C++ classes to scripts.
During some tests to check if there are memory leaks while using this classes in scripts, I figured out, that destructors of those exposed classes never get called.

if I write a script like:
```python
for x in range(0, 50000):
	obj = ClassObject()
	obj.textattribute = "Hello"
	print("Created object.")
```
I would expect that python cleans up the objects at some time.
If my class has a destructor with some log outputs, I can see, that the destructor never gets called.

Am I doing something wrong, or is this supposed to be the case, that exposed meta classes destructors aren't registered in script interfaces ?

Regards

**Comment by wqking** [comment]
The C++ object created by the script should be freed when the script object is freed (GCed), so the destructor should be called.
How did you register the meta data, and what Python do you use?
I just took the unit test and verified the destructor is called in Python 2.7.

**Comment by wqking** [comment]
I checked your implementation and can't see obvious mistake. However, your implementation seems unnecessary complicated and why you can't load cpgf core to script and you need `ScriptMysqlBindHelper`?  

> it was really hard to figure out how to use cpgf just with all the small samples inside cpgf  

Did you check the [documentation on cpgf.org](http://www.cpgf.org/document/index.html)?

I suggest you start a refresh test application to nail down the problem.  
My [2D game engine Gincu](https://github.com/wqking/gincu) always used the script binding so you may reference it. It's much simpler than your implementation.  
The entry is [runScriptMain() in scriptmain.cpp](https://github.com/wqking/gincu/blob/master/tests/application/script/scriptmain.cpp), it uses [Gincu gscriptingmain.cpp](https://github.com/wqking/gincu/blob/master/src/scripting/gscriptingmain.cpp), and the core is [cpgf script runner](https://github.com/cpgf/cpgf/blob/master/src/scriptbind/gscriptrunner.cpp).  
The meta data is generated from the tool metagen. If you write the meta data manually, the trick is to put the meta data in a virtual name space and reject that name space to the script, so you don't need to bind the classes/functions manually.  
The Lua scripts is in [this folder](https://github.com/wqking/gincu/tree/master/bin/resources/lua)

Hope that helps.

**Comment by ErrorFlexXx** [comment]
Thank you for your reply.

The cpgf core is loaded now - I simply executed the script code before binding the core, so it wasn't available :+1: But, I think I will continue using my helper class for the MYSQL_BIND structure, because it's simpler to use on script side than dealing with pointer to binary arrays to make the mysql C-Api happy.

Thanks for the link to Gincu, I adopted the template function for invoking script functions with variadic paramterlists. Really nice.

One problem left, the uncalled destructor. I tested a bit with my code and read the docs.
The docs state: 
`
GDefineMetaClass::define = GDefineMetaClass::declare + add to global
GDefineMetaClass::declare will create a meta class, and GDefineMetaClass owns the meta class. That means GDefineMetaClass will free the meta class in its destructor.
`
Does that mean, if I GDefineMetaClass::define (like I've done for all classes) the object won't get freed ?

I'll setup a small project to test it with the original script runner.

Regards

**Comment by wqking** [comment]
Your quote document has nothing to do with your problem. It's how cpgf free the **meta class**, not any script object.
I set breakpoint at `~TestObject` in testscriptbindmetadata2.h, then in any unit test script I just create an instance of TestObject.
Let me know your test result on the small project.

**Comment by ErrorFlexXx** [comment]
Hello again,

I tried to build up a small project, but unfortunately I can't get it simpler than my implementation in OpenGMP. The scriptloaders aren't compiled for some reason, so the linker always complains about unresolved functions. Maybe this was the reason, why I implemented my own script loader (Nevertheless I need my own implementations anyway, to get unique scopes in python for each script).

Anyway, I took the unit test, too and put outputs in the destructor of the TestObject, and it gets called. But I'm not sure, if the test objects are generated from a loaded script or from a c++ test case setup function, which would be another situation.

I would have liked to compile another cpgf project, like gincu and create objects from scripts there and observe, if destructors gets called or not, but I didn't manage to compile gincu on linux (it's ok, the readme states that linux wasn't tested).

Do you have e.g. a working gincu compilation and can do that test ?
So by just adding a loop in a script that creates ~100.000 objects and look at the ram, if the objects are destroyed later on ?

**Comment by wqking** [comment]
I may have reproduced it in Gincu. I will investigate on it.

**Comment by ErrorFlexXx** [comment]
Thank you for this. If I can help with some tests or anything, just let me know.

**Comment by wqking** [comment]
The issue should be caused by commit cff244e215261ad593e544de99dafd5d1f862ec7, which optimized the performance by reusing the metatable, which causes the Lua binding engine holds strong reference to the C++ script object and prevents it from GCed.  
It's similar in Python binding.  
It's not trivial fix so I need more time. You can either wait for the update or get older commits (but which may have problem in Python binding).

**Comment by wqking** [comment]
Fixed this issue on Lua engine, a32aa03fc16ce81d95e0d1f90dba39c39a62aa1f  
Added unit tests for such kind of memory leak.  
Python engine needs extra work, I will fix it later.

**Comment by ErrorFlexXx** [comment]
Just tested this update - It works as expected.
No leaks, everythink is fine :+1: 
Thank you for this.

**Comment by wqking** [comment]
I added object leak test for Python (d180c92780545454d91a2cc73a374d8181e3987d) and it works.  
So Python engine should already work now without any fixing.  
Unlike Lua, Python engine doesn't have the bug that the cached object holds strong reference, so the Python engine doesn't need to fix. That Python leak bug was caused by that in the common script binding code the internal object pool holds strong reference to objects which prevents the objects being destroyed, and it was already fixed.
